pkg/runtime/scheduler/scheduler.go

```diff
@@ -0,0 +1,287 @@
+						if err := stream.CloseSend(); err != nil {
+							log.Debugf("Error closing stream")
+						}
+						if nerr := client.closeConnection(); nerr != nil {
+							log.Debugf("Error closing connection: %v", nerr)
+						}
+						return streamerr
+					}
+				}
+				log.Infof("Received response: %+v %+v", resp.GetData(), resp.GetMetadata()) // TODO rm this once it sends the triggered job back to the app
+			}
+		}
+	}
+}
+
+func (m *Manager) establishSchedulerConn(ctx context.Context, client *schedulerClient, streamReq *schedulerv1pb.StreamJobRequest) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			// Keep trying to establish connection and process stream
+			err := m.establishConnection(ctx, client)
+			if err != nil {
+				log.Debugf("Error establishing connection: %v", err)
+				continue
+			}
+			err = m.processStream(ctx, client, streamReq)
+			if err != nil {
+				log.Debugf("Error processing stream: %v", err)
+				continue
+			}
+		}
+	}
+}
+
+// watchJob starts watching for job triggers from a single scheduler client.
+func (m *Manager) watchJob(ctx context.Context, client *schedulerClient) error {
+	streamReq := &schedulerv1pb.StreamJobRequest{
+		AppId:     m.connDetails.AppID,
+		Namespace: m.connDetails.Namespace,
+	}
+	return m.establishSchedulerConn(ctx, client, streamReq)
+}
+
+func (client *schedulerClient) closeConnection() error {
+	if client.conn == nil {
+		return nil // Connection is already closed
+	}
+
+	if err := client.conn.Close(); err != nil {
+		return fmt.Errorf("error closing connection: %v", err)
+	}
+
+	client.conn = nil // Set the connection to nil to indicate it's closed
+	return nil
+}
+
+// closeAndReconnect closes the connection and reconnects the client.
+func (client *schedulerClient) closeAndReconnect(ctx context.Context) error {
+	if client.conn != nil {
+		if err := client.conn.Close(); err != nil {
+			return fmt.Errorf("error closing connection: %v", err)
+		}
+	}
+	log.Info("Attempting to connect to Scheduler")
+	conn, schedulerClient, err := schedulerclient.New(ctx, client.address, client.security)
+	if err != nil {
+		return fmt.Errorf("error creating Scheduler client for address %s: %v", client.address, err)
+	}
+	client.conn = conn
+	client.scheduler = schedulerClient
+
+	log.Infof("Reconnected to scheduler at address %s", client.address)
+	return nil
+}
+
+// NextClient returns the next client in a round-robin manner.
+func (m *Manager) NextClient() schedulerv1pb.SchedulerClient {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	// Check if there is only one client available
+	if len(m.clients) == 1 {
+		return m.clients[0].scheduler
+	}
+
+	// Randomly choose the starting index && get the next client
+	//nolint:gosec // there is no need for a crypto secure rand.
+	startingIndex := rand.Intn(len(m.clients))
+	nextIdx := (m.lastUsedIdx + startingIndex) % len(m.clients)
+	nextClient := m.clients[nextIdx]
```

Comment:
This logic seems too complicated to simply pick one from the index. Why not simply `nextIdx := rand.Intn(len(m.clients))` ?